service/remotecfg/remotecfg.go

```diff
@@ -0,0 +1,318 @@
+	return fmt.Sprintf("%x", fnvHash.Sum(nil))
+}
+
+// Service implements a service for remote configuration.
+// The default value of ch is nil; this means it will block forever if the
+// remotecfg service is not configured. In addition, we're keeping track of
+// the ticker so we can avoid leaking goroutines.
+// The datapath field is where the service looks for the local cache location.
+// It is defined as a hash of the Arguments field.
+type Service struct {
+	opts Options
+	args Arguments
+
+	ctrl service.Controller
+
+	mut               sync.RWMutex
+	asClient          agentv1connect.AgentServiceClient
+	ch                <-chan time.Time
+	ticker            *time.Ticker
+	dataPath          string
+	currentConfigHash string
+}
+
+// ServiceName defines the name used for the remotecfg service.
+const ServiceName = "remotecfg"
+
+// Options are used to configure the remotecfg service. Options are
+// constant for the lifetime of the remotecfg service.
+type Options struct {
+	Logger      log.Logger // Where to send logs.
+	StoragePath string     // Where to cache configuration on-disk.
+}
+
+// Arguments holds runtime settings for the remotecfg service.
+type Arguments struct {
+	URL              string                   `river:"url,attr,optional"`
+	ID               string                   `river:"id,attr,optional"`
+	Metadata         map[string]string        `river:"metadata,attr,optional"`
+	PollFrequency    time.Duration            `river:"poll_frequency,attr,optional"`
+	HTTPClientConfig *config.HTTPClientConfig `river:",squash"`
+}
+
+// GetDefaultArguments populates the default values for the Arguments struct.
+func GetDefaultArguments() Arguments {
+	return Arguments{
+		ID:               agentseed.Get().UID,
+		Metadata:         make(map[string]string),
+		PollFrequency:    1 * time.Minute,
+		HTTPClientConfig: config.CloneDefaultHTTPClientConfig(),
+	}
+}
+
+// SetToDefault implements river.Defaulter.
+func (a *Arguments) SetToDefault() {
+	*a = GetDefaultArguments()
+}
+
+// Validate implements river.Validator.
+func (a *Arguments) Validate() error {
+	// We must explicitly Validate because HTTPClientConfig is squashed and it
+	// won't run otherwise
+	if a.HTTPClientConfig != nil {
+		return a.HTTPClientConfig.Validate()
+	}
+
+	return nil
+}
+
+// Hash marshals the Arguments and returns a hash representation.
+func (a *Arguments) Hash() (string, error) {
+	b, err := river.Marshal(a)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal arguments: %w", err)
+	}
+	return getHash(b), nil
+}
+
+// New returns a new instance of the remotecfg service.
+func New(opts Options) (*Service, error) {
+	basePath := filepath.Join(opts.StoragePath, ServiceName)
+	err := os.MkdirAll(basePath, 0750)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		opts:   opts,
+		ticker: time.NewTicker(math.MaxInt64),
+	}, nil
+}
+
+// Data is a no-op for the remotecfg service.
+func (s *Service) Data() any {
+	return nil
```

Comment:
Should this be a panic instead? Normally we dont panic but this feels like the same category as duplicate component ids.

Reply:
I don't mind either way, but I don't see a benefit for panicking here.